### PR TITLE
Dynatrace metrics exporter

### PR DESCRIPTION
**Pull request by dyladan**
**Description:** Add an exporter for metrics to Dynatrace

**Link to tracking Issue:** n/a

**Testing:** Tests are done for configuration and some serialization

**Documentation:** README.md added in `exporter/dynatraceexporter` which explains configuration and use of Dynatrace metric exporter.

### CODE REVIEW

**Comment by dyladan** [comment]
@kbrockhoff not sure who to tag here but you're assigned to this PR so I guess you're as good a person to ask as any. The tests seem to have failed in a completely unrelated part of the codebase (xray receiver). Is there a way you can force a rerun? Or should I just push another commit to trigger the build?

File: exporter/dynatraceexporter/serialization/serialization.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	output := ""
+	for i := 0; i < data.Len(); i++ {
+		p := data.At(i)
+		if p.IsNil() {
+			continue
+		}
+
+		tagline := serializeTags(p.LabelsMap(), tags)
+		count := float64(p.Count())
+		avg := p.Sum() / count
```
Comment (anuraaga) [inline exporter/dynatraceexporter/serialization/serialization.go:80]: Think we need to handle `count == 0`

File: exporter/dynatraceexporter/serialization/serialization.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+	output := ""
+	for i := 0; i < data.Len(); i++ {
+		p := data.At(i)
+		if p.IsNil() {
+			continue
+		}
+
+		tagline := serializeTags(p.LabelsMap(), tags)
+
+		avg := float64(p.Sum()) / float64(p.Count())
```
Comment (anuraaga) [inline exporter/dynatraceexporter/serialization/serialization.go:103]: Ditto

File: exporter/dynatraceexporter/serialization/serialization.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+func serializeTags(labels pdata.StringMap, exporterTags []string) string {
+	tags := append([]string{}, exporterTags...)
+	labels.ForEach(func(k string, v string) {
+		key, err := NormalizeString(strings.ToLower(k), maxDimKeyLen)
+		if err != nil {
+			return
+		}
+		tag := key + "=" + v
```
Comment (anuraaga) [inline exporter/dynatraceexporter/serialization/serialization.go:134]: Assuming it's safe to not have any escaping or anything of the value?

File: exporter/dynatraceexporter/metrics_exporter.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+func (e *exporter) send(message string) error {
+	e.logger.Debug("Sending lines to Dynatrace\n" + message)
+	req, err := http.NewRequest("POST", e.cfg.API.URL, bytes.NewBufferString(message))
+	if err != nil {
+		return fmt.Errorf("dynatrace error while creating HTTP request: %s", err.Error())
+	}
+
+	req.Header.Add("Content-Type", "text/plain; charset=UTF-8")
+
+	if len(e.cfg.API.APIToken) != 0 {
```
[the remark this answers is not in the document]
Comment (dyladan) [reply exporter/dynatraceexporter/metrics_exporter.go:124]: I moved the validation while developing. you are correct it's validated

File: exporter/dynatraceexporter/serialization/serialization.go (hunk 3 above)
Comment (dyladan) [reply exporter/dynatraceexporter/serialization/serialization.go]: Added escaping but I think the characters requiring escaping are not allowed by the spec anyways

**Comment by dyladan** [comment]
@anuraaga I addressed all of your comments except I am still building the string. I will try to get around to refactoring it to use a buffer but i don't know when that will be done and I dont think it should block this PR.

File: exporter/dynatraceexporter/metrics_exporter.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+			return 0, fmt.Errorf("error processing data:, %s", err.Error())
+		}
+	}
+
+	return dropped, nil
+}
+
+func (e *exporter) send(message string) error {
+	e.logger.Debug("Sending lines to Dynatrace\n" + message)
+	req, err := http.NewRequest("POST", e.cfg.API.URL, bytes.NewBufferString(message))
```
[the remark this answers is not in the document]
Comment (dyladan) [reply exporter/dynatraceexporter/metrics_exporter.go]: I put this on my todo list but I don't think it should block this PR.

File: exporter/dynatraceexporter/LICENSE [hunk 6]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
+                           Version 2.0, January 2004
+                        http://www.apache.org/licenses/
+
+   TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION
```
Comment (bogdandrutu) [inline exporter/dynatraceexporter/LICENSE:5]: Why is this needed?

File: exporter/dynatraceexporter/config/config.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+	"go.opentelemetry.io/collector/config/configmodels"
+)
+
+// Config defines configuration for the Dynatrace exporter.
+type Config struct {
+	configmodels.ExporterSettings `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct.
+
+	// URL of the Dynatrace metrics ingest endpoint. Defaults to http://127.0.0.1:14499/metrics/ingest
+	Endpoint string `mapstructure:"endpoint"`
```
Comment (bogdandrutu) [inline exporter/dynatraceexporter/config/config.go:29]: Do you want to use the confighttp.HttpClientSettigs from core?

**Review by bogdandrutu — COMMENTED** [review]
Please follow the instructions https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#when-adding-a-new-component

File: exporter/dynatraceexporter/factory.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	exp := newMetricsExporter(params, cfg)
+
+	return exporterhelper.NewMetricsExporter(
+		cfg,
+		params.Logger,
+		exp.PushMetricsData,
+		exporterhelper.WithQueue(exporterhelper.CreateDefaultQueueSettings()),
+		exporterhelper.WithRetry(exporterhelper.CreateDefaultRetrySettings()),
```
Comment (kbrockhoff) [inline exporter/dynatraceexporter/factory.go:76]: Do you have plans to make these configurable in the future?

File: exporter/dynatraceexporter/metrics_exporter.go [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+			continue
+		}
+		libraryMetrics := resourceMetric.InstrumentationLibraryMetrics()
+		for j := 0; j < libraryMetrics.Len(); j++ {
+			libraryMetric := libraryMetrics.At(j)
+			if libraryMetric.IsNil() {
+				continue
+			}
+			metrics := libraryMetric.Metrics()
+			for k := 0; k < metrics.Len(); k++ {
```
Comment (kbrockhoff) [inline exporter/dynatraceexporter/metrics_exporter.go:70]: Consider handling processing for loop in a separate method.

File: exporter/dynatraceexporter/config/config.go (hunk 7 above)
Comment (dyladan) [reply exporter/dynatraceexporter/config/config.go:29]: I wasn't aware it existed. I just looked at another exporter and did what it did. I think it was datadog? I can change to use that if you would prefer?

File: exporter/dynatraceexporter/LICENSE (hunk 6 above)
Comment (dyladan) [reply exporter/dynatraceexporter/LICENSE:5]: I'll remove it, it was added by my client and I didn't notice.

**Comment by dyladan** [comment]
> Please follow the instructions [open-telemetry/opentelemetry-collector@`master`/CONTRIBUTING.md#when-adding-a-new-component](https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#when-adding-a-new-component)

So would you like me to open a new PR with just the Readme, configuration, and factory?

File: exporter/dynatraceexporter/factory.go (hunk 8 above)
Comment (dyladan) [reply exporter/dynatraceexporter/factory.go:76]: No. This is just what I saw another exporter do.

Comment (kbrockhoff) [reply exporter/dynatraceexporter/factory.go:81]: This is fine to start but as this gets deployed to different environments you will likely find this needs to be adjustable.

**Review by kbrockhoff — APPROVED** [review]

**Comment by dyladan** [comment]
What is the best way for me to resolve go.sum issues? Should I just re-run `go build` or should I use `go mod tidy`? `go mod tidy` removes the `memcachedreceiver` also for some reason.

File: exporter/dynatraceexporter/metrics_exporter.go (hunk 9 above)
Comment (dyladan) [reply exporter/dynatraceexporter/metrics_exporter.go:72]: All that is done in the loop is some nil checking then it delegates to the correct serialization function. I could make the whole for loop body its own function but I think it unnecessary redirection while reading the code.

**Comment by dyladan** [comment]
I just wanted to check in to see what the current status of this is. I don't think this PR is waiting on me to do something but if it is please let me know. I'm not sure how many reviews are needed for this repo before a merge, but I think I've addressed all of the open PR comments. I have also removed the component from the components.go file as suggested by bogdan to limit the impact of merging this component until its stability is proven.

**Comment by tigrannajaryan** [comment]
@bogdandrutu do you have any other comments or good to merge?

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**